src/mesh_analysis.rs

```diff
@@ -0,0 +1,413 @@
+            n( 0.0,  1.0,  0.0),
+            n( 0.0,  1.0,  0.0),
+            n( 0.0,  1.0,  0.0),
+            // front
+            n( 0.0, -1.0,  0.0),
+            n( 0.0, -1.0,  0.0),
+            n( 0.0, -1.0,  0.0),
+            n( 0.0, -1.0,  0.0),
+            // top
+            n( 0.0,  0.0,  1.0),
+            n( 0.0,  0.0,  1.0),
+            n( 0.0,  0.0,  1.0),
+            n( 0.0,  0.0,  1.0),
+            // bottom
+            n( 0.0,  0.0, -1.0),
+            n( 0.0,  0.0, -1.0),
+            n( 0.0,  0.0, -1.0),
+            n( 0.0,  0.0, -1.0),
+            // right
+            n( 1.0,  0.0,  0.0),
+            n( 1.0,  0.0,  0.0),
+            n( 1.0,  0.0,  0.0),
+            n( 1.0,  0.0,  0.0),
+            // left
+            n(-1.0,  0.0,  0.0),
+            n(-1.0,  0.0,  0.0),
+            n(-1.0,  0.0,  0.0),
+            n(-1.0,  0.0,  0.0),
+        ];
+
+        #[rustfmt::skip]
+            let faces = vec![
+            // back
+            TriangleFace::new(2, 1, 0),
+            TriangleFace::new(2, 3, 0),
+            // top
+            TriangleFace::new(2, 1, 5),
+            TriangleFace::new(2, 5, 6),
+            // right
+            TriangleFace::new(2, 6, 7),
+            TriangleFace::new(7, 3, 2),
+            // bottom
+            TriangleFace::new(3, 7, 4),
+            TriangleFace::new(4, 0, 3),
+            // front
+            TriangleFace::new(6, 4, 7),
+            TriangleFace::new(4, 6, 5),
+            // left
+            TriangleFace::new(0, 4, 5),
+            TriangleFace::new(5, 1, 0),
+        ];
+
+        Geometry::from_triangle_faces_with_vertices_and_normals(
+            faces,
+            vertex_positions,
+            vertex_normals,
+        )
+    }
+
+    #[test]
+    fn test_mesh_analysis_find_edge_with_valency() {
+        let (faces, vertices) = quad();
+        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+            faces.clone(),
+            vertices.clone(),
+            NormalStrategy::Sharp,
+        );
+        let oriented_edges: Vec<OrientedEdge> = geometry.oriented_edges_iter().collect();
+        let edge_valency_map = edge_valencies(&oriented_edges);
+
+        let oriented_edges_with_valency_1_correct = vec![
+            OrientedEdge::new(0, 1),
+            OrientedEdge::new(1, 2),
+            OrientedEdge::new(2, 3),
+            OrientedEdge::new(3, 0),
+        ];
+        let oriented_edges_with_valency_2_correct =
+            vec![OrientedEdge::new(2, 0), OrientedEdge::new(0, 2)];
+
+        let oriented_edges_with_valency_1 = find_edges_with_valency(&edge_valency_map, 1);
+        let oriented_edges_with_valency_2 = find_edges_with_valency(&edge_valency_map, 2);
+
+        for o_e in oriented_edges_with_valency_1 {
+            assert!(oriented_edges_with_valency_1_correct
+                .iter()
+                .any(|e| *e == o_e));
+        }
+
+        for o_e in oriented_edges_with_valency_2 {
```

Comment:
If you iterate in this order, the test doesn't prove anything. If `find_edges_with_valency()` returned `(0, 2)` 2 times, this test would still pass but the logic would be incorrect. Instead the `_correct` data should be in the outer loop.